artifacts/definitions/Windows/Forensics/Usn.yaml

```diff
@@ -12,41 +15,82 @@ description: |
 
-parameters:
-  - name: PathRegex
-    description: A regex to match the entire path (you can watch a directory or a file type).
-    default: .
-    type: regex
+type: CLIENT
 
+parameters:
   - name: Device
     description: The NTFS drive to parse
     default: C:\\
-
   - name: AllDrives
     description: Dump USN from all drives and VSC
     type: bool
-
-precondition: SELECT OS from info() where OS = "windows"
-
+  - name: PathRegex
+    description: A regex to match the entire path (you can watch a directory or a file type).
+    default: .
+    type: regex
+  - name: MFT_ID_Regex
+    description: A regex to match the MFTId. e.g ^10225$ or ^(10225|232111)$
+    default: .
+    type: regex
+  - name: Parent_MFT_ID_Regex
+    description: A regex to match the MFTId. e.g ^10225$ or ^(10225|232111)$
+    default: .
+    type: regex
+  - name: DateAfter
+    type: timestamp
+    description: "search for events after this date. YYYY-MM-DDTmm:hh:ssZ"
+  - name: DateBefore
+    type: timestamp
+    description: "search for events before this date. YYYY-MM-DDTmm:hh:ssZ"
 sources:
-  - query: |
-      LET all_drives = SELECT
-        OSPath.Components[0] AS Drive
-      FROM glob(globs="/*/$Extend/$UsnJrnl:$J", accessor="ntfs")
-      WHERE log(message="Processing " + Drive)
+  - precondition:
+      SELECT OS From info() where OS =~ 'windows'
+
+    query: |
+      -- firstly set timebounds for performance
+      LET DateAfterTime <= if(condition=DateAfter,
+            then=timestamp(epoch=DateAfter), else=timestamp(epoch="1600-01-01"))
+      LET DateBeforeTime <= if(condition=DateBefore,
+            then=timestamp(epoch=DateBefore), else=timestamp(epoch="2200-01-01"))
+        
+      LET all_drives = SELECT OSPath.Components[0] AS Drive
+        FROM glob(globs="/*/$Extend/$UsnJrnl:$J", accessor="ntfs")
+        WHERE log(message="Processing " + Drive)
 
-      SELECT * FROM if(condition=AllDrives,
-      then={
-        SELECT * FROM foreach(row=all_drives,
-        query={
-          SELECT *, Drive AS Device
-          FROM parse_usn(
-             device=Drive, accessor="ntfs")
-          WHERE FullPath =~ PathRegex
-        })
-      }, else={
-        SELECT *, Device
-        FROM parse_usn(device=Device, accessor="ntfs")
-        WHERE FullPath =~ PathRegex
-      })
+      SELECT 
+            Timestamp,
+            Filename,
+            Device,FullPath,
+            Reason,
+            _FileMFTID as MFTId,
+            _FileMFTSequence as Sequence,
+            _ParentMFTID as ParentMFTId,
+            _ParentMFTSequence as ParentSequence,
+            FileAttributes,
+            SourceInfo,
+            Usn
+      FROM if(condition=AllDrives,
+          then={
+            SELECT * FROM foreach(row=all_drives,
+            query={
+              SELECT *, Drive AS Device
+              FROM parse_usn(
+                 device=Drive, accessor="ntfs")
+              WHERE FullPath =~ PathRegex
```

Comment:
As an optimization place FullPath at the end because it is a dynamic function and will not run unless it is used. It is actually quite expensive to run.